cl/sentinel/gossip.go

```diff
@@ -285,30 +289,55 @@ type GossipSubscription struct {
 	host         peer.ID
 	ch           chan *GossipMessage
 	ctx          context.Context
+	expiration   atomic.Value // Unix nano for how much we should listen to this topic
+	subscribed   atomic.Bool
 
 	topic *pubsub.Topic
 	sub   *pubsub.Subscription
 
 	cf context.CancelFunc
 	rf pubsub.RelayCancelFunc
 
-	setup  sync.Once
 	stopCh chan struct{}
 }
 
-func (sub *GossipSubscription) Listen() (err error) {
-	sub.setup.Do(func() {
-		sub.stopCh = make(chan struct{}, 3)
-		sub.sub, err = sub.topic.Subscribe()
-		if err != nil {
-			err = fmt.Errorf("failed to begin topic %s subscription, err=%w", sub.topic.String(), err)
-			return
+func (sub *GossipSubscription) Listen() {
+	go func() {
```

Comment:
I feel check it every 100ms is a bit cpu wasting.
how do you feel following pseudo code? I try to rewrite it with no short time checking, but still weird i think.
just try if can inspire you other ideas

```golang
func checkStopSub() {
	if isSubscribe {
		// close something
		isSubscribe = false
	}
}
func checkStartSub() {
	if !isSubscribe {
		sub = retrySubscribeUntilSuccess()
		isSubscribe = true
	}
}
go func() {
  sub := retrySubscribeUntilSuccess()
  defer sub.Close()
  for {
  	timer := time.After(expirationTime.Sub(time.Now()))
        select{
        case <-ctx.Done():
           return
        case <-timer:
	   checkStopSub()
	   expirationTime = forever
	case exp := <-expTimCh:
          // deal with new expiration time
	  if exp.Before(time.Now()) {
	     checkStopSub()
	     expirationTime = forever
	  } else {
	    checkStartSub()
	    expirationTime = exp
	  }
       }
       timer.Stop()
   }
}
```

Reply:
100ms is a low for a human, not for a computer, a computer execute a clock cycle every nanosecond